File: src/mtglab/claude/prices.py
"""What the tokens cost, in money.

**This module reverses a decision that was written down.** `mtglab claude usage`
reported tokens and not dollars, deliberately, and said why:

    prices move (Sonnet 5's introductory rate ends 2026-08-31) and a stale
    hardcoded price table would turn an honest count into a wrong invoice

That objection was right and is not withdrawn. What changed (2026-08-18,
Aaron's call) is that "how much is this instance costing" became a question
worth answering imperfectly rather than not at all -- an app that is free
forever is one whose bill somebody has to actually watch. So the table exists,
and every part of its design is an answer to the sentence above rather than a
way around it.

**Rates are dated, and the date is rendered.** `CHECKED` is when a human last
read the pricing page, and every surface that shows a figure shows that date
next to it. A number whose age is visible is a number a reader can discount;
one that merely looks current is the wrong invoice the old comment feared.

**A rate that is known to move is modelled, not flattened.** Sonnet 5's
introductory pricing ends 2026-08-31, and that is the whole reason the objection
was phrased the way it was -- so a table that pretended it did not would fail
on its own headline example within a fortnight of being written. A model may
carry a `Window`: a rate, and the date after which a different one applies.
The estimate picks by the conversation's own timestamp, so a bill spanning the
changeover is priced correctly on both sides of it.

**A model the table does not know is priced at nothing and counted.** The
ledger stores the *served-by* id, which can be a model this build never heard
of -- a fallback, a newer id, an A/B via `MTGLAB_CLAUDE_MODEL`. Guessing a rate
for it would be the wrong invoice; charging it zero without saying so would be
worse, because it would read as "cheap". Every caller gets `unpriced` beside
the figure and is expected to render it.

**The figure stays a floor, for the reason it always was.** Prompt cache
*writes* bill at 1.25x input and are recorded nowhere (schema v7 says so in the
column comments), so no arithmetic here can see them. Cache *reads* are
captured and are priced at a tenth of input, which is the number that justifies
the cache existing.

Rates are per million tokens, US dollars, from the Anthropic pricing page.
Re-check them by reading that page and editing this file; there is no fetch,
deliberately -- a price this project quotes should change when a person decided
it should, not when a request happened to succeed.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date
from typing import Any

#: When a human last read the pricing page. Rendered beside every figure this
#: module produces, and the only honest substitute for a freshness guarantee --
#: the same instrument ADR 19 gave the dossier.
CHECKED = date(2026, 8, 18)

#: Where to re-check. Not fetched; see the module docstring.
SOURCE = "https://platform.claude.com/docs/en/pricing"


@dataclass(frozen=True)
class Rate:
    """Dollars per million tokens, input and output."""

    input: float
    output: float

    #: Prompt cache reads, as a fraction of the input rate. One number rather
    #: than a third column because it is the same ratio across the family and
    #: a per-model copy of it would be three chances to mistype a tenth.
    CACHE_READ_FRACTION = 0.1


@dataclass(frozen=True)
class Priced:
    """One model's rate, and the date any of it is known to change."""

    rate: Rate
    #: A rate that takes over `after`. `None` when nothing is scheduled --
    #: which is most models, and is not a promise that nothing will change.
    then: Rate | None = None
    #: The last day `rate` applies. Required when `then` is set.
    until: date | None = None

    def __post_init__(self) -> None:
        if (self.then is None) != (self.until is None):
            raise ValueError(
                "a scheduled rate change needs both `then` and `until` -- one "
                "without the other prices half a changeover and silently "
                "keeps the wrong side of it")

    def on(self, when: date) -> Rate:
        """The rate in force on `when`."""
        if self.until is not None and self.then is not None and when > self.until:
            return self.then
        return self.rate


# Checked 2026-08-18 against the pricing page. Sonnet 5 is the one with a
# scheduled change, and it is the example the old no-price-table comment named:
# $2/$10 introductory through 2026-08-31, $3/$15 after. Modelling it is the
# point -- see the module docstring.
PRICES: dict[str, Priced] = {
    "claude-fable-5": Priced(Rate(input=10.00, output=50.00)),
    "claude-mythos-5": Priced(Rate(input=10.00, output=50.00)),
    "claude-opus-5": Priced(Rate(input=5.00, output=25.00)),
    "claude-opus-4-8": Priced(Rate(input=5.00, output=25.00)),
    "claude-opus-4-7": Priced(Rate(input=5.00, output=25.00)),
    "claude-opus-4-6": Priced(Rate(input=5.00, output=25.00)),
    "claude-sonnet-5": Priced(
        Rate(input=2.00, output=10.00),
        then=Rate(input=3.00, output=15.00),
        until=date(2026, 8, 31),
    ),
    "claude-sonnet-4-6": Priced(Rate(input=3.00, output=15.00)),
    "claude-haiku-4-5": Priced(Rate(input=1.00, output=5.00)),
}

_PER_MILLION = 1_000_000


@dataclass
class Estimate:
    """What a set of ledger rows came to, and what could not be priced.

    `unpriced_models` is not a footnote. A row whose model this table does not
    know contributes nothing to `usd`, so a caller that renders the figure
    without rendering this is showing a number that is wrong downward and
    reads as reassuring.
    """

    usd: float = 0.0
    #: Conversations whose model carried no rate.
    unpriced: int = 0
    #: Which ids those were, so somebody can go and add them.
    unpriced_models: set[str] = field(default_factory=set)

    @property
    def complete(self) -> bool:
        """Whether every row counted was priced."""
        return self.unpriced == 0

    def as_dict(self) -> dict[str, Any]:
        return {
            "usd": round(self.usd, 4),
            "unpriced": self.unpriced,
            "unpriced_models": sorted(self.unpriced_models),
            "complete": self.complete,
            "checked": CHECKED.isoformat(),
        }


def rate_for(model: str, when: date | None = None) -> Rate | None:
    """The rate for `model` on `when`, or `None` if this table has never
    heard of it. `when` defaults to today."""
    priced = PRICES.get(model)
    if priced is None:
        return None
    return priced.on(when or date.today())


def cost(*, model: str, input_tokens: int, output_tokens: int,
         cache_read_tokens: int = 0, when: date | None = None) -> float | None:
    """What one conversation cost, or `None` for a model with no rate.

    `None` rather than `0.0`, and the distinction is the whole reason this
    returns an optional: a conversation that cost nothing and a conversation
    nobody can price are different facts, and a caller that cannot tell them
    apart will report the second as the first.

    Cache reads are priced separately at a tenth of input. They are counted
    *beside* `input_tokens` and never inside it, because the API reports
    `input_tokens` as the uncached remainder only -- adding them would
    double-count the cheap half of every cached prompt.
    """
    rate = rate_for(model, when)
    if rate is None:
        return None
    return (
        input_tokens * rate.input
        + output_tokens * rate.output
        + cache_read_tokens * rate.input * Rate.CACHE_READ_FRACTION
    ) / _PER_MILLION


def estimate(rows: list[dict[str, Any]], *,
             when: date | None = None) -> Estimate:
    """Total `rows`, each carrying `model` and the three token counts.

    Rows are the ledger's own shape (`ledger.summary(by='model')` and its
    per-mode sibling), so this works over any grouping that kept the model.
    A row with no `model` key at all is unpriceable by definition and counts
    as one, rather than raising -- a roll-up that lost the axis is a caller
    bug, and the honest report of it is a figure that says it is incomplete.
    """
    out = Estimate()
    for row in rows:
        model = str(row.get("model") or "")
        conversations = int(row.get("conversations") or 1)
        amount = cost(
            model=model,
            input_tokens=int(row.get("input_tokens") or 0),
            output_tokens=int(row.get("output_tokens") or 0),
            cache_read_tokens=int(row.get("cache_read_tokens") or 0),
            when=when,
        )
        if amount is None:
            out.unpriced += conversations
            out.unpriced_models.add(model or "(none recorded)")
            continue
        out.usd += amount
    return out


def render(amount: float) -> str:
    """A dollar figure a person can read, down to the tenth of a cent.

    Sub-cent totals are the ordinary case on an instance this size, and
    `$0.00` beside a real conversation reads as "nothing happened" rather than
    as "not very much" — which is the wrong lesson to draw from a bill that is
    working. Four decimals until a cent, two after.
    """
    if amount and abs(amount) < 0.01:
        return f"${amount:.4f}"
    return f"${amount:,.2f}"

File: go/internal/prices/prices.go
// Package prices is the Claude ledger's price table: what the tokens cost,
// in money.
//
// The standing rules, all load-bearing: rates are dated and the date
// renders beside every figure; a rate known to move is modelled as a window
// rather than flattened; **a window is priced at the rates that were in force
// during it**, not at this morning's; a model the table does not know is
// priced at nothing and **counted**, because charging it zero silently would
// read as "cheap"; and the figure stays a floor, since cache writes bill at
// 1.25x input and are recorded nowhere. Re-check rates by reading the pricing
// page and editing `Table`; the corpus test holds the table to the recorded
// rates, so a rate never drifts silently.
//
// The third rule is the one that was broken by a caller rather than by this
// file, which is why `Segments` lives here now. `Over` prices a set of rows at
// one date, and every caller passed `Today()` — correct until the day a rate
// moved, then wrong forever for everything spent before it, by exactly the
// size of the move. Sonnet 5's introductory window closed on 2026-08-31 and
// the whole recorded bill before it started reading ~50% high the next
// morning. A caller with a date range asks `Segments` to cut it at the
// boundaries the table itself knows, prices each piece at the rate that was in
// force, and adds them with `Sum`.
package prices

import (
	"sort"
	"time"

	"github.com/aasquier/sylvan-library/go/internal/floats"
	"github.com/aasquier/sylvan-library/go/internal/wire"
)

// Checked is when a human last read the pricing page — rendered beside every
// figure, the only honest substitute for a freshness guarantee.
const Checked = "2026-08-18"

// Source is where to re-check. Not fetched, deliberately.
const Source = "https://platform.claude.com/docs/en/pricing"

// CacheReadFraction prices a prompt-cache read as a fraction of input. One
// number rather than a per-model column: it is the same ratio across the
// family, and a copy per model would be that many chances to mistype a tenth.
const CacheReadFraction = 0.1

// Rate is dollars per million tokens.
type Rate struct {
	Input, Output float64
}

// Priced is one model's rate, and the date any of it is known to change:
// `Then` takes over after `Until`, both set or neither.
type Priced struct {
	Rate  Rate
	Then  *Rate
	Until string // ISO date, the last day Rate applies
}

// On is the rate in force on `when` (an ISO date; string comparison is date
// comparison).
func (p Priced) On(when string) Rate {
	if p.Until != "" && p.Then != nil && when > p.Until {
		return *p.Then
	}
	return p.Rate
}

// Table is the rate per model. Sonnet 5 is the one with a scheduled
// change: $2/$10 introductory through 2026-08-31, $3/$15 after.
var Table = map[string]Priced{
	"claude-fable-5":    {Rate: Rate{10.00, 50.00}},
	"claude-mythos-5":   {Rate: Rate{10.00, 50.00}},
	"claude-opus-5":     {Rate: Rate{5.00, 25.00}},
	"claude-opus-4-8":   {Rate: Rate{5.00, 25.00}},
	"claude-opus-4-7":   {Rate: Rate{5.00, 25.00}},
	"claude-opus-4-6":   {Rate: Rate{5.00, 25.00}},
	"claude-sonnet-5":   {Rate: Rate{2.00, 10.00}, Then: &Rate{3.00, 15.00}, Until: "2026-08-31"},
	"claude-sonnet-4-6": {Rate: Rate{3.00, 15.00}},
	"claude-haiku-4-5":  {Rate: Rate{1.00, 5.00}},
}

const perMillion = 1_000_000

// Cost is what one conversation came to, or false for a model with no rate
// — a conversation that cost nothing and one nobody can price are different
// facts. Every product is fenced with `floats.Rounded`, because `a*b + c`
// is the shape arm64 fuses into one FMADDD and the recorded arithmetic
// rounds after every product.
func Cost(model string, inputTokens, outputTokens, cacheReadTokens int64,
	when string) (float64, bool) {
	priced, ok := Table[model]
	if !ok {
		return 0, false
	}
	rate := priced.On(when)
	total := floats.Rounded(float64(inputTokens) * rate.Input)
	total = floats.Rounded(total + floats.Rounded(float64(outputTokens)*rate.Output))
	cache := floats.Rounded(floats.Rounded(float64(cacheReadTokens)*rate.Input) * CacheReadFraction)
	total = floats.Rounded(total + cache)
	return total / perMillion, true
}

// Row is one ledger summary row's pricing-relevant half.
type Row struct {
	Model         string
	Conversations int64
	InputTokens   int64
	OutputTokens  int64
	CacheRead     int64
}

// Estimate is the total, and what could not be
// priced — which is not a footnote, since an unpriced row contributes
// nothing to `usd` and a caller rendering the figure without it is showing a
// number that is wrong downward.
type Estimate struct {
	USD            float64
	Unpriced       int64
	UnpricedModels []string // sorted, "(none recorded)" for a blank model
}

// Over totals rows in order — a plain running sum, never a compensated
// one, deliberately: the recorded totals accumulate term by term, and the
// order of additions is part of the answer's last bit.
func Over(rows []Row, when string) Estimate {
	out := Estimate{}
	seen := map[string]bool{}
	for _, row := range rows {
		conversations := row.Conversations
		if conversations == 0 {
			conversations = 1
		}
		amount, ok := Cost(row.Model, row.InputTokens, row.OutputTokens,
			row.CacheRead, when)
		if !ok {
			out.Unpriced += conversations
			name := row.Model
			if name == "" {
				name = "(none recorded)"
			}
			if !seen[name] {
				seen[name] = true
				out.UnpricedModels = append(out.UnpricedModels, name)
			}
			continue
		}
		out.USD = floats.Rounded(out.USD + amount)
	}
	sort.Strings(out.UnpricedModels)
	return out
}

// AsDict is the wire shape, in the recorded key order: usd (rounded
// half-to-even to four places, rendered canonically), unpriced,
// unpriced_models, complete, checked.
func (e Estimate) AsDict() wire.OrderedMap {
	models := e.UnpricedModels
	if models == nil {
		models = []string{}
	}
	return wire.OrderedMap{
		{Key: "usd", Value: floats.Float(floats.RoundTo(e.USD, 4))},
		{Key: "unpriced", Value: e.Unpriced},
		{Key: "unpriced_models", Value: models},
		{Key: "complete", Value: e.Unpriced == 0},
		{Key: "checked", Value: Checked},
	}
}

// Today is the local day as an ISO date — on the instance, UTC.
func Today() string { return time.Now().Format("2006-01-02") }

// ---- pricing a window that crosses a rate change --------------------------

// Segment is one stretch of time over which every rate in Table held still:
// `[Since, Until)`, and `When` is a date inside it, which is all `Cost` needs
// to pick the rate that was in force.
//
// Both bounds are compared as text against `created_at`, which is ISO-8601
// UTC — so a bare date is a perfectly good instant here, and midnight is where
// it lands. An empty `Since` is "from the beginning" and an empty `Until` is
// "up to now"; both are the absence of a bound rather than a bound at zero.
type Segment struct {
	Since string
	Until string
	When  string
}

// Boundaries are the dates the table's rates change, sorted, deduplicated —
// the first day each new rate applies, which is the day AFTER the `Until` a
// Priced records as the last day of the old one.
//
// Derived from Table rather than written down beside it, because a second
// list of the same dates is a second thing to forget when a rate moves.
func Boundaries() []string {
	seen := map[string]bool{}
	out := []string{}
	for _, priced := range Table {
		if priced.Until == "" || priced.Then == nil {
			continue
		}
		day, err := time.Parse("2006-01-02", priced.Until)
		if err != nil {
			// An unparseable Until cannot be turned into a boundary, and
			// guessing one would price a window at the wrong rate silently.
			// `Priced.On` still compares it as text, so the table keeps working;
			// what is lost is only the split, which is the honest failure.
			continue
		}
		changeover := day.AddDate(0, 0, 1).Format("2006-01-02")
		if !seen[changeover] {
			seen[changeover] = true
			out = append(out, changeover)
		}
	}
	sort.Strings(out)
	return out
}

// Segments splits `[since, now]` at every boundary it crosses.
//
// **This is what "estimated spend" means.** Pricing a whole window at today's
// rate is not an estimate of what was spent, it is an estimate of what the
// same traffic would cost if it were bought again this morning — and when a
// rate moves by half, as Sonnet 5's introductory window did on 2026-09-01,
// the two answers differ by half. The ledger records when each conversation
// happened; this is what lets a roll-up use that.
//
// `now` bounds the split rather than the sum: a boundary that has not arrived
// yet cannot have any rows past it, so a segment beginning after today is left
// out instead of costing a query that can only answer zero.
func Segments(since, now string) []Segment {
	cuts := []string{since}
	for _, boundary := range Boundaries() {
		// `>` and not `>=`: a boundary at exactly the window's start has
		// already taken effect, so there is nothing before it to price.
		if boundary > since && boundary <= now {
			cuts = append(cuts, boundary)
		}
	}
	out := make([]Segment, 0, len(cuts))
	for i, start := range cuts {
		seg := Segment{Since: start, When: now}
		if i+1 < len(cuts) {
			seg.Until = cuts[i+1]
			// A date inside the segment: the day before the next change, which
			// is the last day this segment's rates were in force. Any date in
			// the range would pick the same rate; this one always exists.
			seg.When = dayBefore(seg.Until)
		}
		out = append(out, seg)
	}
	return out
}

func dayBefore(date string) string {
	day, err := time.Parse("2006-01-02", date)
	if err != nil {
		return date
	}
	return day.AddDate(0, 0, -1).Format("2006-01-02")
}

// Sum folds one estimate per segment into the window's own.
//
// The dollars are added in segment order with `Over`'s rounding, so the sum is
// the same arithmetic done in the same sequence rather than a second kind of
// addition. The unpriced counts add; the unpriced model names are a set, since
// a model the table cannot price is unpriceable in every segment it appears in
// and naming it three times would read as three problems.
func Sum(parts []Estimate) Estimate {
	out := Estimate{}
	seen := map[string]bool{}
	for _, part := range parts {
		out.USD = floats.Rounded(out.USD + part.USD)
		out.Unpriced += part.Unpriced
		for _, model := range part.UnpricedModels {
			if !seen[model] {
				seen[model] = true
				out.UnpricedModels = append(out.UnpricedModels, model)
			}
		}
	}
	sort.Strings(out.UnpricedModels)
	return out
}
